### ISSUE TITLE
Number of images via v3 API

### ISSUE THREAD

**Issue by lschr** [imageio/imageio#915]
Is there a way to get the number of images in a file via the v3 API?

**Comment by FirefoxMetzger** [comment]
The first thing that comes to mind is something like

```python
import imageio.v3 as iio

n_images = iio.improps(ImageResource, index=...).shape[0] 
```

This works if you have multiple images of the same shape, e.g.:

```python
iio.improps("imageio:chelsea.png", index=...).shape[0]  # 1
iio.improps("imageio:newtonscradle.gif", index=...).shape[0]  # 36
iio.improps("imageio:cockatoo.mp4", index=..., plugin="pyav").shape[0]   # 280
```

However, if you have multiple images with different shapes (especially inside a TIFF) then your options are more limited. Here, you can either inspect the format-specific metadata (via `immeta`), or you can count via

```python
n_images = len([x for x in iio.imiter(ImageResource)])
```

Neither approach is ideal though, because immeta doesn't report the number of contained images for all formats and counting can be expensive as it decodes pixel data.

I think it could be useful to add a `n_images` field to `ImageProperties` to create a uniform way of checking this. What do you think?

---

Along those lines, it would probably also be useful to have a `format` and `backend` field that gives you a programmatic way of inspecting the file format (e.g. if you are reading bytes) and the backend (e.g., to check which one was automatically selected) ...

**Comment by lschr** [comment]
> Here, you can either inspect the format-specific metadata (via `immeta`), or you can count via
> 
> ```python
> n_images = len([x for x in iio.imiter(ImageResource)])
> ```

This would mean reading potentially gigabytes of data just to get the number of frames, so I think it is not really an option.

> I think it could be useful to add a `n_images` field to `ImageProperties` to create a uniform way of checking this. What do you think?

I think this is a great idea. For TIFF, it would be possible to return either the number of series (e.g. if `index=..., page=None`) or the number of pages (e.g. `index=..., page=...` or `index=None, page=...`).

**Comment by lschr** [comment]
I started implementing this, but then realized that it would not work so well. Returning the number of images only makes sense when querying file-global properties (i.e. `index=None`), but not when querying metadata for a single image.

What would you think about adding a new method, e.g. `frame_count()`, to the `PluginV3` class, which is to be implemented by subclasses?

**Comment by FirefoxMetzger** [comment]
> I started implementing this, but then realized that it would not work so well. Returning the number of images only makes sense when querying file-global properties (i.e. index=None), but not when querying metadata for a single image.

What is the argument against `n_images` being `1` if the user calls improps with `index!=...`?

> What would you think about adding a new method, e.g. frame_count(), to the PluginV3 class, which is to be implemented by subclasses?

Hm, what does this add that `improps(...).n_images` doesn't do?

**Comment by lschr** [comment]
I would say the number of images is a property of the file, not of an image. It seems strange to me to have a property `n_images` associated with each image, which is set to 1, since it isn't really applicable.

From a programming POV it is no problem to set `n_images = 1` if index is `None` (in case index is `...` it should probably also be 1). If you like it this way, I can go ahead with it.

**Comment by lschr** [comment]
> Hm, what does this add that `improps(...).n_images` doesn't do?

Some plugins read an image (in case of `index=...` even all images) to get `shape` and/or `dtype`, which is not necessary to get the number of images. Thus, having a separate method would be beneficial for performance.

**Comment by lschr** [comment]
> FYI: The only plugins that read image/pixel data to provide `props` are the v2 plugins via their legacy wrapper. These exist for backward compatibility because the old API wasn't designed to separate pixel data and metadata. We now have v3 plugins for pillow, pyav, and tifffile (added this year tada) which makes `props` cheap for all the commonly used formats.

`OpenCVPlugin.properties` reads pixel data to find out the image shape. If index is `...`, it even reads all images (although this could be changed easily to read only one image). The same is true for `LegacyPlugin.properties`.

It seems like v3 plugins prepend the number of images to `ImageProperties.shape`, which actually makes `ImageProperties.n_images` redundant… So, maybe it is not necessary after all? If having a separate method for getting the number of images is out of the question, wouldn't it be sufficient to have more efficient implementations of `OpenCVPlugin.properties` and `LegacyPlugin.properties` plus modifying `TifffilePlugin.properties` to return the number of pages instead of the number of series as `shape[0]` in case `page=...`?

**Comment by lschr** [comment]
> It is slightly tangential to this issue, but I was thinking to introduce `props.extension = ".tiff"` and `props.plugin = "tifffile"` to the list of props to make it easier to discover which plugin was selected to read a resource and what container one is dealing with.

This would be very useful (for me at least).

**Comment by FirefoxMetzger** [comment]
> OpenCVPlugin.properties reads pixel data to find out the image shape.

Good catch. I forgot about the OpenCV backend. I haven't yet found a way to extract metadata (including `n_images`) using OpenCV that doesn't involve parsing pixel data, so our options are either to parse the pixel data and extract `props` from the result or fail. In this situation, I think it is better to do what is expected of us and be slow than to fail at something that could be done. 

If you know a better way to get metadata from OpenCV I'm very happy to improve it :) 

That said, I don't think there are any formats that are only covered by OpenCV and nothing else. Typically, either pillow, pyav, or tifffile will also support the format if OpenCV supports it. The only time (I think) you'd prefer OpenCV is for performance reasons because it's faster at writing PNGs than any of the other backends. 

> The same is true for LegacyPlugin.properties.

Yes, those are the v2 plugins. In order to not lose support for any container while we migrate from v2 plugins to v3 plugins we have `LegacyPlugin` which wraps the old `format`s and implements the new API on top of them. I guess we would have an angle for reading `n_images` without touching pixel data here if we'd use `Reader.get_length()`, but I don't think we can get other information (shape/dtype) this way. It's also something I hope to make disappear in the coming years as we port more backends from `Format` to `PluginV3`.

Is this why you suggested to introduce a dedicated `frame_count()` method?

> It seems like v3 plugins prepend the number of images to ImageProperties.shape, which actually makes ImageProperties.n_images redundant… So, maybe it is not necessary after all?

Yes and no. `ImageProperties.shape` will include a batch dimension if called with `index=...`. This is the currently suggested way to read the number of images, but it's erroneous if the individual images don't have compatible shapes. If a container stores multiple images with different shapes (which is super rare in practice), then `ImageProperties.shape` will report `(n_images, *shape_of_first_image)` in most cases. I would much prefer to report `ImageProperties.shape=None` instead (since `shape` is undefined in this case) and use `ImageProperties.n_images` as a canonical source of the number of images in a resource, which works irrespective of the individual shapes.

So far the current behavior of `ImageProperties.shape` has not caused any reported issues, so I have been allocating my time to things that are more pressing (like a PluginV3 version of tifffile). Nonetheless, it is something that should be addressed, which is why I think something like `n_images` is a useful addition.

> [...] plus modifying TifffilePlugin.properties to return the number of pages instead of the number of series as shape[0] in case page=...?

You mean that calling `improps(..., plugin="tifffile", index=..., page=...).shape` should return `(n_pages, *page_shape)` assuming that all the pages have the same shape? 

I think we could make this happen. It would mean that we need to modify `TifffilePlugin.read` to match that behavior, i.e. `TifffilePlugin.read(..., index=..., page=...)` would read all pages, stack them into a batch, and return the result, but that too is possible.

**Comment by lschr** [comment]
This can be closed, right?
